### ISSUE TITLE
Add support for .MENU files

### ISSUE THREAD

**Issue by thebeardedgeek** [codefori/vscode-ibmi#1049]
I am working on building menus on my IBM i. I wanted to avoid using SDA and was introduced to UIM menus (thanks @martintosney for showing me this). 

I would like to see if we can add support for the .MENU object type. This would include:

- syntax highlighting if possible
- compile options

Here is an example of a compile command:
`CRTMNU MENU(&OPENLIB/&OPENMBR) TYPE(*UIM) SRCFILE(&OPENLIB/&OPENSPF) SRCMBR(&OPENMBR) INCFILE(UTIL/QPNLREF) REPLACE(*YES)`

The INCFILE provides import files for this command. Martin provided some examples here:
https://github.com/martintosney/uim-menu

I can provide an example of the .MENU contents if needed.

**Comment by chrjorgensen** [comment]
Hi @thebeardedgeek 

Great idea - but I think most is already available:
- UIM menus are defined in the UIM language and are panels like other panel types. If you use the source member type `PNLGRP` for the UIM menu source, the syntax will be highlighted by the [vscode-ibmi-languages extension](https://marketplace.visualstudio.com/items?itemName=barrettotte.ibmi-languages), which supports UIM keywords (though some menu keywords are not included in the extension and thus not syntax highlighted, but I will make a PR for this).
- The compile command can be created as a user defined action.

So you have the tools already to be able to create UIM menus. Only the compile command action is not there yet.

**Comment by thebeardedgeek** [comment]
Thanks @chrjorgensen. I added the compile command already and I will look at the highlighting. I was thinking more for the benefit of others that this be included. I appreciate your response.

**Comment by worksofliam** [comment]
So to me it sounds like we need to 

* add a file type association for `.menu` for the `pnlgrp` language id.
* add a new default compile command

**Comment by thebeardedgeek** [comment]
Here is what syntax highlighting looks like when I choose PNLGRP:
![image](https://user-images.githubusercontent.com/9466407/218118933-443aa068-e63b-4adb-aa49-f4d9a040b35e.png)

I am also included an example of a menu to have this all in one place:
[EXAMPLE.zip](https://github.com/halcyon-tech/vscode-ibmi/files/10708391/EXAMPLE.zip)

**Comment by chrjorgensen** [comment]
@thebeardedgeek The syntax highlighting will work when [PR #121](https://github.com/barrettotte/vscode-ibmi-languages/pull/121) in the vscode-ibmi-languages has been merged by @barrettotte. I just spent a few hours creating and testing this PR.

IMHO I don't think we need (yet) another source type `.menu` - why not just use the existing type `.pnlgrp`? Only thing needed is an action CRTMNU for the `.pnlgrp` source type.

Btw. the default action CRTPNLGRP seems to only be defined for members, not streamfiles. This should be fixed as well.

😃

**Comment by chrjorgensen** [comment]
@thebeardedgeek FYI, this is how your example source looks like with [PR #121](https://github.com/barrettotte/vscode-ibmi-languages/pull/121) applied:

![billede](https://user-images.githubusercontent.com/13275072/218213736-87652c00-686b-4b32-9003-9c7a34c2a476.png)

**Comment by thebeardedgeek** [comment]
@chrjorgensen I used .MENU based on the fact that IBM has them listed that way on their menu files. Here is an example:
![image](https://user-images.githubusercontent.com/9466407/218221383-6d839e22-1f98-4d78-a5a6-4006af384588.png)

I am not sure what is the recommended, most common, or typical way that people will create menu files but I made the suggestion to add .MENU so that the Code for IBM i extension covers these objects.

**Comment by chrjorgensen** [comment]
@thebeardedgeek I understand what you're thinking...

...but I think you're mixing OBJECT types (like *MENU, *FILE etc) and MEMBER types (like RPGLE, DSPF etc):
There is no 1to1 relationship between object types and member (=source) type. Multiple member types can compile into the same object type, e.g. RPGLE, CLLE, CLE will all compile into object type *PGM. And one member type can compile into multiple object types, e.g. RPGLE can compile into *MODULE, *PGM and *SRVPGM based on the compile command used.

VS Code and Code for IBM i is mostly about developing and compiling source into objects. So the focus is on the source (=member) type, not the object type. We do not handle the *PGM object type (except for the object browser), instead we handle the source types RPGLE, CLLE etc.

The source type is just for linking to the components / commands handling the source - and anyone can invent their own source types (you use MENU and Martin Tosney uses UIM). But the components / commands are built for one or more specific types and will generally only support the predefined types. The Code for IBM i extension allows for some circumvention by user defined actions, but the IBM i languages extension providing the syntax highlighting is built for the predefined types. You _can_ manually select the PNLGRP source type for a source with type MENU to get the correct syntax highlighting, but it will not be selected automatically (unless the extension is changed to handle your MENU source type as well). 

As you have already created an action to run CRTMNU, you just need to alter it to (also) work for PNLGRP source types. Then you have two commands for the PNLGRP source types - CRTMNU and CRTPNLGRP - and you select the command for the object you want to create - menu or panel group. If you choose the wrong command, the compiler will tell you! 😄 

So my advice will be that you stick with the predefined source type PNLGRP and use your action for CRTMNU on this source type.

As for what is used to create menus by the IBM i community, I don't know - but based on what I've seen through my years on this platform I guess very few IBM i professionals actually know the UIM language and use it for menus. I think the majority either use a display file (together with a message file) or a program...

**Comment by thebeardedgeek** [comment]
@chrjorgensen I see what you are saying. I took the steps you mentioned and I have syntax highlighting and I can compile the member if I tie PNLGRP to CRTMNU. If I had known about UIM back when I was creating more screens I think I would have opted for this route instead of the hours it would take to create a DSPF, PGM, etc. Each have their place and value though.

Thank you for the info.

**Comment by worksofliam** [comment]
@thebeardedgeek I am still interested in adding a `CRTMNU` default since we don't have one.

Would you mind sharing the Action you have setup?

**Comment by worksofliam** [comment]
@chrjorgensen Thank you the PR you created over on IBMi Languages. I think that will make @thebeardedgeek and Martin's experience just a tad better!

**Comment by thebeardedgeek** [comment]
This is what I am using but it can be modified. The INCFILE might be different for others but I have my menu import files in the same library as these new menus I am building.
```
{
"type": "member",
"extensions": [
"MENU",
"PNLGRP"
],
"environment": "ile",
"name": "Create Menu (CRTMNU)",
"command": "?CRTMNU MENU(&OPENLIB/&OPENMBR) TYPE(*UIM) SRCFILE(&OPENLIB/&OPENSPF) SRCMBR(&OPENMBR) INCFILE(&OPENLIB/QPNLREF) REPLACE(*YES)"
 }
```

**Comment by chrjorgensen** [comment]
@thebeardedgeek I'm very curious about the import files (I have never used it myself). Does the :IMPORT tag copy other members from the INCFILE source file? It's not documented in the [Application Display Programming](https://www.ibm.com/support/knowledgecenter/ssw_ibm_i_61/rzakc/sc415715.pdf) manual, which only talks about other PNLGRP objects...?

**Comment by chrjorgensen** [comment]
@thebeardedgeek Just wanted to let you know, that the new UIM menu tags support is included in the newest [release 0.6.9](https://github.com/barrettotte/vscode-ibmi-languages/releases/tag/v0.6.9) of the [IBM i Languages extension](https://github.com/barrettotte/vscode-ibmi-languages), so your menu sources will have syntax highlighting after upgrade.

Hope this helps!

**Comment by thebeardedgeek** [comment]
@chrjorgensen yes, you can create files with settings that can be imported using the INCFILE. This allows for them to be used across multiple UIM/MENU programs.

@martintosney sent me this article which helped with understanding UIM menus:
https://www.mcpressonline.com/programming-other/general/create-ibmstyle-rolling-menus-with-uim

Is it possible to add the .UIM file type to IBMi Languages? It shows as Type: Plain Text currently. If I change it to PNLGRP it looks good.
![image](https://user-images.githubusercontent.com/9466407/218551972-44bce01a-4cb5-4d25-9930-997c507451e2.png)

**Comment by chrjorgensen** [comment]
@thebeardedgeek 

> yes, you can create files with settings that can be imported using the INCFILE. This allows for them to be used across multiple UIM/MENU programs. Does that answer your question?

Yes, thank you. The documentation is just not clear on this. And thanks for the link to the previous century... 1993, just five years after the birth of the AS/400! 😆 

You can setup file association for UIM files: Open a UIM file, click on the file type (which you've marked in your screenshot) and configure a file association:

![billede](https://user-images.githubusercontent.com/13275072/218576064-5c729ecf-89d0-4848-be83-644577b209b9.png)

Select PNLGRP as the association:

![billede](https://user-images.githubusercontent.com/13275072/218576819-2a92fee2-5498-4479-bc01-6bfdd9ebb528.png)

From now on your UIM files will be considered a PNLGRP file, including syntax highlighting.
File association is a standard feature of VS Code and not specific to any file type.

**Comment by thebeardedgeek** [comment]
LOL...sorry the documentation on this is not very available but I thought that article had some halfway decent info. The import files remind me of some of the things we have available now to centralize commonly used code.

Thank you for the information on File Associations. I went ahead and did what you recommended.

**Comment by chrjorgensen** [comment]
@thebeardedgeek Just wanted to inform you that I've finalized support for all UIM tags in PNLGRP and also the `.im` (imbed) directive, and it is included in the latest version [0.6.10](https://github.com/barrettotte/vscode-ibmi-languages/releases/tag/0.6.10) of [vscode-ibmi-languages](https://github.com/barrettotte/vscode-ibmi-languages) extension.

So all your UIM menu tags and imbeds should be properly syntax highlighted now.
